File: ash-zig/src/maintenance.zig
const std = @import("std");

pub const Command = struct {
    name: []const u8,
    description: []const u8,
    cmd: []const u8,
    args: []const []const u8,
    requires_sudo: bool,
    useful: bool,
};

pub const CommandResult = struct {
    command: *const Command,
    success: bool,
    output_buf: [4096]u8 = undefined,
    output_len: usize = 0,
    error_msg_buf: [256]u8 = undefined,
    error_msg_len: usize = 0,
    duration_ns: u64 = 0,

    pub fn output(self: *const CommandResult) []const u8 {
        return self.output_buf[0..self.output_len];
    }

    pub fn errorMsg(self: *const CommandResult) ?[]const u8 {
        if (self.error_msg_len == 0) return null;
        return self.error_msg_buf[0..self.error_msg_len];
    }
};

/// Available maintenance commands
pub const commands = [_]Command{
    .{
        .name = "Flush DNS Cache",
        .description = "Clear DNS resolver cache",
        .cmd = "dscacheutil",
        .args = &.{"-flushcache"},
        .requires_sudo = false,
        .useful = true,
    },
    .{
        .name = "Restart mDNSResponder",
        .description = "Restart multicast DNS service",
        .cmd = "killall",
        .args = &.{ "-HUP", "mDNSResponder" },
        .requires_sudo = true,
        .useful = true,
    },
    .{
        .name = "Rebuild Spotlight Index",
        .description = "Reindex Spotlight search database",
        .cmd = "mdutil",
        .args = &.{ "-E", "/" },
        .requires_sudo = true,
        .useful = true,
    },
    .{
        .name = "Rebuild Launch Services",
        .description = "Rebuild app registration database",
        .cmd = "/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister",
        .args = &.{ "-kill", "-r", "-domain", "local", "-domain", "user" },
        .requires_sudo = false,
        .useful = true,
    },
    .{
        .name = "Clear Font Cache",
        .description = "Remove cached font data",
        .cmd = "atsutil",
        .args = &.{ "databases", "-remove" },
        .requires_sudo = true,
        .useful = true,
    },
    .{
        .name = "Purge RAM",
        .description = "Free inactive memory (limited benefit)",
        .cmd = "purge",
        .args = &.{},
        .requires_sudo = true,
        .useful = false,
    },
};

/// Run a single maintenance command
pub fn run(allocator: std.mem.Allocator, cmd: *const Command) !CommandResult {
    const start = std.time.nanoTimestamp();

    var result = CommandResult{
        .command = cmd,
        .success = false,
    };

    // Build argv
    var argv = std.ArrayList([]const u8){};
    defer argv.deinit(allocator);

    if (cmd.requires_sudo) {
        try argv.append(allocator, "sudo");
    }
    try argv.append(allocator, cmd.cmd);
    for (cmd.args) |arg| {
        try argv.append(allocator, arg);
    }

    // Execute
    var child = std.process.Child.init(argv.items, allocator);
    child.stdout_behavior = .Pipe;
    child.stderr_behavior = .Pipe;

    try child.spawn();

    // Read output
    const stdout = child.stdout orelse return result;
    const output = stdout.readToEndAlloc(allocator, 4096) catch "";
    defer if (output.len > 0) allocator.free(output);

    const stderr = child.stderr orelse return result;
    const err_output = stderr.readToEndAlloc(allocator, 256) catch "";
    defer if (err_output.len > 0) allocator.free(err_output);

    const term = try child.wait();

    const end = std.time.nanoTimestamp();
    result.duration_ns = @intCast(end - start);

    // Copy output
    if (output.len > 0) {
        const len = @min(output.len, result.output_buf.len);
        @memcpy(result.output_buf[0..len], output[0..len]);
        result.output_len = len;
    }

    if (err_output.len > 0) {
        const len = @min(err_output.len, result.error_msg_buf.len);
        @memcpy(result.error_msg_buf[0..len], err_output[0..len]);
        result.error_msg_len = len;
    }

    result.success = term.Exited == 0;

    return result;
}

/// Run all useful commands
pub fn runAll(allocator: std.mem.Allocator, useful_only: bool) !std.ArrayList(CommandResult) {
    var results = std.ArrayList(CommandResult){};
    errdefer results.deinit(allocator);

    for (&commands) |*cmd| {
        if (useful_only and !cmd.useful) continue;
        const result = try run(allocator, cmd);
        try results.append(allocator, result);
    }

    return results;
}

/// Check if a command is available
pub fn isCommandAvailable(allocator: std.mem.Allocator, cmd_name: []const u8) bool {
    var child = std.process.Child.init(&.{ "which", cmd_name }, allocator);
    child.stdout_behavior = .Ignore;
    child.stderr_behavior = .Ignore;

    child.spawn() catch return false;
    const term = child.wait() catch return false;

    return term.Exited == 0;
}

// Tests
test "commands defined" {
    try std.testing.expect(commands.len > 0);
    try std.testing.expectEqualStrings("Flush DNS Cache", commands[0].name);
}

File: crates/ash-sdk/src/maintenance.rs
use std::process::Command;
use std::time::Instant;

use serde::{Deserialize, Serialize};

use crate::error::{AshError, ErrorCode, Result};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MaintenanceCommand {
    pub name: String,
    pub description: String,
    pub requires_sudo: bool,
    pub useful: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MaintenanceCatalog {
    pub commands: Vec<MaintenanceCommand>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MaintenanceRunRequest {
    pub name: String,
    pub dry_run: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct MaintenanceCommandResult {
    pub name: String,
    pub dry_run: bool,
    pub success: bool,
    pub output: String,
    pub elapsed_ms: u64,
}

pub fn list_maintenance_commands() -> MaintenanceCatalog {
    MaintenanceCatalog {
        commands: vec![
            MaintenanceCommand {
                name: "dns.flush".to_string(),
                description: "Flush resolver caches and restart mDNSResponder.".to_string(),
                requires_sudo: true,
                useful: true,
            },
            MaintenanceCommand {
                name: "launchservices.rebuild".to_string(),
                description: "Rebuild the Launch Services registration database.".to_string(),
                requires_sudo: false,
                useful: true,
            },
            MaintenanceCommand {
                name: "spotlight.reindex".to_string(),
                description: "Reindex Spotlight from the filesystem root.".to_string(),
                requires_sudo: true,
                useful: true,
            },
            MaintenanceCommand {
                name: "font-cache.clear".to_string(),
                description: "Remove cached font databases.".to_string(),
                requires_sudo: true,
                useful: true,
            },
        ],
    }
}

pub fn run_maintenance_command(request: MaintenanceRunRequest) -> Result<MaintenanceCommandResult> {
    let start = Instant::now();
    let requires_sudo = list_maintenance_commands()
        .commands
        .into_iter()
        .find(|command| command.name == request.name)
        .map(|command| command.requires_sudo)
        .ok_or_else(|| {
            AshError::new(
                ErrorCode::Unsupported,
                format!("unknown maintenance command: {}", request.name),
                "run `ash maintenance list --json` to inspect supported maintenance commands",
            )
        })?;
    if request.dry_run {
        return Ok(MaintenanceCommandResult {
            elapsed_ms: start.elapsed().as_millis() as u64,
            dry_run: true,
            name: request.name,
            output: "dry-run: no command executed".to_string(),
            success: true,
        });
    }
    if requires_sudo && !effective_user_is_root()? {
        return Err(AshError::new(
            ErrorCode::PrerequisiteBlocked,
            format!(
                "maintenance command {} requires root privileges",
                request.name
            ),
            "re-run the command from a root shell or keep using --dry-run",
        ));
    }

    let output = match request.name.as_str() {
        "dns.flush" => {
            let first = Command::new("/usr/bin/dscacheutil")
                .arg("-flushcache")
                .output()
                .map_err(|error| {
                    AshError::new(
                        ErrorCode::MaintenanceFailed,
                        format!("failed to run dscacheutil: {error}"),
                        "ensure the command exists and retry from a normal macOS shell",
                    )
                })?;
            if !first.status.success() {
                return Err(AshError::new(
                    ErrorCode::MaintenanceFailed,
                    "dscacheutil flushcache failed",
                    String::from_utf8_lossy(&first.stderr).trim().to_string(),
                ));
            }
            Command::new("/usr/bin/killall")
                .args(["-HUP", "mDNSResponder"])
                .output()
                .map_err(|error| {
                    AshError::new(
                        ErrorCode::MaintenanceFailed,
                        format!("failed to restart mDNSResponder: {error}"),
                        "retry from a root shell",
                    )
                })?
        }
        "launchservices.rebuild" => Command::new(
            "/System/Library/Frameworks/CoreServices.framework/Frameworks/LaunchServices.framework/Support/lsregister",
        )
        .args(["-kill", "-r", "-domain", "local", "-domain", "user"])
        .output()
        .map_err(|error| {
            AshError::new(
                ErrorCode::MaintenanceFailed,
                format!("failed to rebuild Launch Services: {error}"),
                "verify the lsregister tool exists on this macOS version",
            )
        })?,
        "spotlight.reindex" => Command::new("/usr/bin/mdutil")
            .args(["-E", "/"])
            .output()
            .map_err(|error| {
                AshError::new(
                    ErrorCode::MaintenanceFailed,
                    format!("failed to start Spotlight reindex: {error}"),
                    "retry from a root shell",
                )
            })?,
        "font-cache.clear" => Command::new("/usr/bin/atsutil")
            .args(["databases", "-remove"])
            .output()
            .map_err(|error| {
                AshError::new(
                    ErrorCode::MaintenanceFailed,
                    format!("failed to clear font cache: {error}"),
                    "retry from a root shell",
                )
            })?,
        _ => unreachable!("validated against registry above"),
    };

    if !output.status.success() {
        return Err(AshError::new(
            ErrorCode::MaintenanceFailed,
            format!("maintenance command {} failed", request.name),
            String::from_utf8_lossy(&output.stderr).trim().to_string(),
        ));
    }

    Ok(MaintenanceCommandResult {
        elapsed_ms: start.elapsed().as_millis() as u64,
        dry_run: false,
        name: request.name,
        output: String::from_utf8_lossy(&output.stdout).trim().to_string(),
        success: true,
    })
}

fn effective_user_is_root() -> Result<bool> {
    let output = Command::new("/usr/bin/id")
        .arg("-u")
        .output()
        .map_err(|error| {
            AshError::new(
                ErrorCode::PrerequisiteBlocked,
                format!("failed to determine effective user id: {error}"),
                "ensure `/usr/bin/id` is available and retry",
            )
        })?;
    if !output.status.success() {
        return Err(AshError::new(
            ErrorCode::PrerequisiteBlocked,
            "failed to determine effective user id",
            String::from_utf8_lossy(&output.stderr).trim().to_string(),
        ));
    }
    Ok(String::from_utf8_lossy(&output.stdout).trim() == "0")
}

#[cfg(test)]
mod tests {
    use super::{MaintenanceRunRequest, list_maintenance_commands, run_maintenance_command};
    use crate::ErrorCode;

    #[test]
    fn list_contains_expected_commands() {
        let catalog = list_maintenance_commands();
        assert!(
            catalog
                .commands
                .iter()
                .any(|command| command.name == "dns.flush")
        );
    }

    #[test]
    fn dry_run_succeeds_without_execution() {
        let result = run_maintenance_command(MaintenanceRunRequest {
            name: "dns.flush".to_string(),
            dry_run: true,
        })
        .expect("dry-run maintenance");
        assert!(result.success);
        assert!(result.dry_run);
    }

    #[test]
    fn sudo_commands_are_blocked_when_not_root() {
        let is_root = super::effective_user_is_root().expect("effective uid");
        if is_root {
            return;
        }
        let error = run_maintenance_command(MaintenanceRunRequest {
            name: "dns.flush".to_string(),
            dry_run: false,
        })
        .expect_err("dns.flush should be blocked when not root");
        assert_eq!(error.code, ErrorCode::PrerequisiteBlocked);
    }
}
